app/src/main/kotlin/pro/qyoga/app/therapist/appointments/core/schedule/CalendarPageModel.kt

```diff
@@ -83,8 +87,26 @@ data class CalendarPageModel(
             return timeMarkRows
         }
 
-        private fun Appointment.fallsIn(wallClockDateTime: LocalDateTime, span: Duration) =
-            this.wallClockDateTime >= wallClockDateTime && this.wallClockDateTime < wallClockDateTime + span
+        private fun Appointment.calendarDateTime(currentUserTimeZone: ZoneId) =
+                this.dateTime.atZone(currentUserTimeZone).toLocalDateTime()
+
+        private fun Appointment.endCalendarDateTime(currentUserTimeZone: ZoneId) =
+                (this.dateTime + this.duration).atZone(currentUserTimeZone).toLocalDateTime()
+
+        private fun Appointment.fallsIn(wallClockDateTime: LocalDateTime, span: Duration, currentUserTimeZone: ZoneId) =
+                (this.calendarDateTime(currentUserTimeZone) >= wallClockDateTime
+                        && this.calendarDateTime(currentUserTimeZone) < wallClockDateTime + span)
+                        || (wallClockDateTime.hour == 0 && wallClockDateTime.minute == 0
+                        && this.calendarDateTime(currentUserTimeZone) < wallClockDateTime
+                        && this.endCalendarDateTime(currentUserTimeZone) >= wallClockDateTime)
+
+        private fun Appointment.startsBeforeMidnight(date: LocalDate, currentUserTimeZone: ZoneId) =
+                this.calendarDateTime(currentUserTimeZone).dayOfMonth != this.endCalendarDateTime(currentUserTimeZone).dayOfMonth
+                        && this.endCalendarDateTime(currentUserTimeZone).minute != 0
+                        && this.calendarDateTime(currentUserTimeZone).dayOfMonth != date.plusDays(1).dayOfMonth
```

Comment:
Это вроде можно чутка упростить
```suggestion
                this.calendarDateTime(currentUserTimeZone).dayOfMonth != this.endCalendarDateTime(currentUserTimeZone).dayOfMonth
                        && this.calendarDateTime(currentUserTimeZone).dayOfMonth < date
```